KHR_materials_anisotropy extension branch. (#1798)
KhronosGroup/glTF: extensions/2.0/Khronos/KHR_materials_anisotropy/README.md

# KHR\_materials\_anisotropy

## Contributors

* Jim Eckerlein, Norbert Nopper, UX3D

## Status

Suggestion

## Dependencies

Written against the glTF 2.0 spec.

## Overview

This extension defines the anisotropic property of a material as observable with brushed metals for instance.
An asymetric specular lobe model is introduced to allow for such phenomena.
The visually distinct feature of that lobe is the elongated appearance of the specular reflection.
For a single punctual light source, the specular reflection will eventually degenerate into a zero width line in the limit,
that is where the material is fully anisotropic, as opposed to be fully isotropic in which case the specular reflecation is radially symmetric.

## Extending Materials

Sample values:

```json
{
    "materials": [
        {
            "extensions": {
                "KHR_materials_anisotropy": {
                    "anisotropy": 0.6,

>                     "anisotropyDirection": [0.0, 1.0, 0.0]

Reviewer: Why do we need a separate anisotropy direction vector? We already define a tangent/bitangent direction and if positive anisotropy is in the tangent direction and negative is in the bitangent direction, isn't that enough?

|-------------------------------|----------------------------------------------------------------------|-----------------------------------|--------------------------------|
|**anisotropy**                 | `number`                                                             | The anisotropy.                   | No, default: `0.0`             |
|**anisotropyTexture**          | [`textureInfo`](/specification/2.0/README.md#reference-textureInfo)  | The anisotropy texture.           | No                             |
|**anisotropyDirection**        | `array`                                                              | The anisotropy direction.         | No, default: `[1.0, 0.0, 0.0]` |
|**anisotropyDirectionTexture** | [`textureInfo`](/specification/2.0/README.md#reference-textureInfo)  | The anisotropy direction texture. | No                             |

## Anisotropy

Two new material properties are introduced: an explicit anisotropy parameter and the direction in which the specular reflection elongates relative to the surface tangents.

> The anisotropy parameter is a dimensionaless number in `[-1, 1]` and forms an injective relation to the roughness distribution along two orthogonal directions, one of which is the direction parameter and the other the result of crossing the direction and the geometric normal.

Reviewer: What space is the direction vector in? Same as the normal map, presumably? And what is the distribution function? Our normal roughness is defined by infinitesimal oblate ellipsoids (Trowbridge-Reitz), so presumably this would map to some more complex ellipsoids? And likewise, how does it interact with the normal roughness parameter?

An anisotropy of `1` means that the specular reflection will elongate along the given direction,
while a value of `-1` will elongate it along the computed orthogonal direction.

| | 0.5 | 0.0 | -0.5 |
| --- | --- | --- | --- |
| `[1.0, 0.0, 0.0]` | ![Fig. 3](figures/fig3.jpg) | ![Fig. 2](figures/fig2.jpg) | ![Fig. 1](figures/fig1.jpg)
| `[1.0, 1.0, 0.0]` | ![Fig. 6](figures/fig6.jpg) | ![Fig. 5](figures/fig5.jpg) | ![Fig. 4](figures/fig4.jpg)

To achieve certain surface finishes, it is possible to define the anisotropy and its direction using a texture.

| Anisotropy | Direction |
|:---:|:---:|
| ![Fig. 7](figures/fig7.jpg) | ![Fig. 8](figures/fig8.jpg) |
| ![Fig. 9](figures/fig9.jpg) | ![Fig. 10](figures/fig10.jpg) |

## Implementation

> While uniform and textured anisotropy are multiplied, uniform and textured direction defintions are mutual exclusive and the latter overrides the former. (While it is surely possible to add both directions in terms of their complex argument, the computional overhead may not justify the additional editorial convenience.)

Reviewer 2: Is there a typo here 'defintions' -> 'definitions'

that is where the material is fully anisotropic, as opposed to be fully isotropic in which case the specular reflection is radially symmetric.

## Extending Materials

Sample values:

```json
{
    "materials": [
        {
            "extensions": {
                "KHR_materials_anisotropy": {
                    "anisotropyStrength": 0.6,
                    "anisotropyDirection": 1.57,
                    "anisotropyTexture": {
                        "index": 0
                    }
                }
            }
        }
    ]
}
```

|                         | Type     | Description               | Required           |
| ----------------------- | -------- | ------------------------- | ------------------ |
| **anisotropyStrength**  | `number` | The anisotropy strength.  | No, default: `0.0` |
| **anisotropyDirection** | `number` | The direction of the anisotropy in tangent, bitangent space, measured in radians counter-clockwise from the tangent. When anisotropyTexture is present, anisotropyDirection provides additional rotation to the vectors in the texture. | No, default: `0.0` |

> | **anisotropyTexture**   | [`textureInfo`](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#reference-textureinfo) | The anisotropy texture. Red and green channels represent the anisotropy direction in [-1, 1] tangent, bitangent space. The magnitude of this vector is multiplied by anisotropyFactor to obtain the anisotropy strength. | No |

Reviewer 3: ```suggestion
| **anisotropyTexture**   | [`textureInfo`](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#reference-textureinfo) | The anisotropy texture. Red and green channels represent the anisotropy direction in [-1, 1] tangent, bitangent space. The vector is rotated by anisotropyDirection, and multiplied by anisotropyStrength, to obtain the final anisotropy direction and strength. | No |
```

uniform vec2 u_AnisotropyDirection;

float anisotropy = u_AnisotropyStrength;
vec2 direction = u_AnisotropyDirection;

#if HAS_ANISOTROPY_MAP
vec2 anisotropyTexture = texture(uv, u_AnisotropyTextureSampler).rg * 2.0 - vec2(1.0);
anisotropy *= length(anisotropyTexture);
direction = mat2(direction.x, -direction.y, direction.y, direction.x) * normalize(anisotropyTexture);
#endif

vec3 anisotropicT = normalize(TBN * vec3(direction, 0.0));
vec3 anisotropicB = normalize(cross(normal_geometric, anisotropicT));

float TdotL = dot(anisotropicT, l);
float BdotL = dot(anisotropicB, l);
float TdotH = dot(anisotropicT, h);
float BdotH = dot(anisotropicB, h);

f_specular += intensity * NdotL * BRDF_specularAnisotropicGGX(f0, f90, alphaRoughness,
    VdotH, NdotL, NdotV, NdotH,
    BdotV, TdotV, TdotL, BdotL, TdotH, BdotH, anisotropy);
```

The anisotropic GGX is defined as follows:

```glsl
vec3 BRDF_specularAnisotropicGGX(vec3 f0, vec3 f90, float alphaRoughness,
    float VdotH, float NdotL, float NdotV, float NdotH, float BdotV, float TdotV,
    float TdotL, float BdotL, float TdotH, float BdotH, float anisotropy)
{
    float at = max(alphaRoughness * (1.0 + anisotropy), 0.00001);

>     float ab = max(alphaRoughness * (1.0 - anisotropy), 0.00001);

Reviewer 4: @abwood Found a reference to this formula, it's published.  I'm not convinced it's the best one to use, but I do want to use one that's published somewhere.  The MaterialX one is looking nice to me, perhaps I'll collect some documentation and propose it at the next meeting.

>     vec3 F = F_Schlick(f0, f90, VdotH);

Reviewer: Schlick is great, but it's still an approximation and therefore not normative. I would pull the definition of D out as the one normative part of the spec, maybe shown a little more mathy like in Burley's original. The visibility term is derived from this distribution being height-correlated, so again, no need for it to be normative.

    float V = V_GGX_anisotropic(NdotL, NdotV, BdotV, TdotV, TdotL, BdotL, anisotropy, at, ab);
    float D = D_GGX_anisotropic(NdotH, TdotH, BdotH, anisotropy, at, ab);

    return F * V * D;
}

float D_GGX_anisotropic(float NdotH, float TdotH, float BdotH, float anisotropy, float at, float ab)
{
    float a2 = at * ab;
    vec3 f = vec3(ab * TdotH, at * BdotH, a2 * NdotH);
    float w2 = a2 / dot(f, f);
    return a2 * w2 * w2 / M_PI;
}

float V_GGX_anisotropic(float NdotL, float NdotV, float BdotV, float TdotV, float TdotL, float BdotL,

>     float anisotropy, float at, float ab)

Reviewer: `anisotropy` is an unused parameter in both these functions - `at` and `ab` already represent its effect.

# KHR\_materials\_anisotropy

## Contributors

* Mike Bond, Adobe, [@miibond](https://github.com/MiiBond)
* Ed Mackey, AGI [@emackey](https://github.com/emackey)
* Alex Wood, AGI [@abwood](https://twitter.com/abwood)
* Nicolas Savva, Autodesk, [@nicolassavva-autodesk](https://github.com/nicolassavva-autodesk)
* Tobias Haeussler, Dassault Systemes [@proog128](https://github.com/proog128)
* Bastian Sdorra, Dassault Systemes [@bsdorra](https://github.com/bsdorra)

> * Don McCurdy, Google [@donrmccurdy](https://twitter.com/donrmccurdy)

Reviewer 3: Minor edit – no longer at Google, and no longer using Twitter actively.

```suggestion
* Don McCurdy, Individual Contributor [@donrmccurdy](https://github.com/donmccurdy)
```
